### DISCUSSION TITLE
mender install reboot not executing

### DISCUSSION THREAD

**Discussion by tirankas** [OE4T/tegra-demo-distro#112] [General]
Hi Guys,
Trying to add an artifact to the rootfs to perform some file cleanup after a successful mender update.
I have one artifact that runs at install leave and one at reboot leave

The install leave artifact runs successfully immediately after install.
However, I dont see any sign of the reboot leave being run.
There are no deployment logs in the /var/lib/mender.
I see both scripts in the /data/mender/scripts/
My sequence is:
mender -install demo-image.mender
mender -commit
reboot

Am I missing anything?

Thanks

**Comment by ichergui** [comment]
Hey @tirankas 

You have to reboot manually the device between the `mender -install` and `mender -commit` 

* Here is the steps 
```
$ mender -install artifact.mender
$ reboot
```
* After device booting

```
$ mender -commit
```

**Reply by tirankas** [reply]
Thanks @ichergui 
I also tried this and it doesn't work either...

**Reply by tirankas** [reply]
According to mender state machine I see reboot comes before commit. As far as I know, the new rootfs artifact will be used only after commit and I need the artifact after reboot to be installed onto the new root...

**Reply by ichergui** [reply]
@tirankas 
Could you please share the logs when you try the steps that I was suggesting ?

**Reply by tirankas** [reply]
Hi @ichergui,
There is no deployment logs in /var/lib/mender.
I did as suggested:
# mender -install artifact-rootfs.mender
INFO[0000] Loaded configuration file: /var/lib/mender/mender.conf
INFO[0000] Loaded configuration file: /etc/mender/mender.conf
INFO[0000] Mender running on partition: /dev/mmcblk0p1
INFO[0000] Start updating from local image file: [./artifact-rootfs.mender]
Installing Artifact of size 1028811264...
INFO[0000] No public key was provided for authenticating the artifact
INFO[0000] Update Module path "/usr/share/mender/modules/v3" could not be opened (open /usr/share/mender/modules/v3: no such file or directory). Update modules will not be available
INFO[0000] Opening device "/dev/mmcblk0p18" for writing
INFO[0000] Native sector size of block device /dev/mmcblk0p18 is 512 bytes. Mender will write in chunks of 1048576 bytes
.............................................................. - 100 %
INFO[0028] All bytes were successfully written to the new partition
INFO[0028] The optimized block-device writer wrote a total of 985 frames, where 972 frames did need to be rewritten (i.e., skipped)
INFO[0028] Wrote 1032179712/1032179712 bytes to the inactive partition
INFO[0028] Enabling partition with new image installed to be a boot candidate: 18
INFO[0028] Executing script: ArtifactInstall_Leave_00
INFO[0029] Collected output (stderr) while running script /var/lib/mender/scripts/ArtifactInstall_Leave_00
Error: No such container: test_container_del

---------- end of script output
Use -commit to update, or -rollback to roll back the update.
At least one payload requested a reboot of the device it updated.
# reboot
.......
# mender -commit
INFO[0000] Loaded configuration file: /var/lib/mender/mender.conf
INFO[0000] Loaded configuration file: /etc/mender/mender.conf
INFO[0000] Mender running on partition: /dev/mmcblk0p18
INFO[0000] Update Module path "/usr/share/mender/modules/v3" could not be opened (open /usr/share/mender/modules/v3: no such file or directory). Update modules will not be available
Committing Artifact...
INFO[0000] Committing update


#journalctl -u mender-client
-- Logs begin at Tue 2021-06-01 09:04:05 UTC, end at Tue 2021-06-01 09:08:14 UTC. --
Jun 01 09:04:13 jetson-nano-emmc-cot systemd[1]: Started Mender OTA update service.
Jun 01 09:04:13 jetson-nano-emmc-cot mender[4877]: time="2021-06-01T09:04:13Z" level=debug msg="Reading Mender configuration from file /var/lib/mender/mender.conf" func=github.com/men>
Jun 01 09:04:13 jetson-nano-emmc-cot mender[4877]: time="2021-06-01T09:04:13Z" level=info msg="Loaded configuration file: /var/lib/mender/mender.conf" func=github.com/mendersoftware/m>
Jun 01 09:04:13 jetson-nano-emmc-cot mender[4877]: time="2021-06-01T09:04:13Z" level=debug msg="Reading Mender configuration from file /etc/mender/mender.conf" func=github.com/menders>
Jun 01 09:04:13 jetson-nano-emmc-cot mender[4877]: time="2021-06-01T09:04:13Z" level=info msg="Loaded configuration file: /etc/mender/mender.conf" func=github.com/mendersoftware/mende>
Jun 01 09:04:13 jetson-nano-emmc-cot mender[4877]: time="2021-06-01T09:04:13Z" level=debug msg="Loaded 2 configuration file(s)" func=github.com/mendersoftware/mender/conf.LoadConfig f>
Jun 01 09:04:13 jetson-nano-emmc-cot mender[4877]: time="2021-06-01T09:04:13Z" level=debug msg="Loaded configuration = &conf.MenderConfig{MenderConfigFromFile:conf.MenderConfigFromFil>
Jun 01 09:04:13 jetson-nano-emmc-cot mender[4877]: time="2021-06-01T09:04:13Z" level=debug msg="Verified configuration = &conf.MenderConfig{MenderConfigFromFile:conf.MenderConfigFromF>
Jun 01 09:04:13 jetson-nano-emmc-cot mender[4877]: time="2021-06-01T09:04:13Z" level=debug msg="Have U-Boot variable: mender_check_saveenv_canary=1" func=github.com/mendersoftware/men>
Jun 01 09:04:13 jetson-nano-emmc-cot mender[4877]: time="2021-06-01T09:04:13Z" level=debug msg="List of U-Boot variables:map[mender_check_saveenv_canary:1]" func=github.com/mendersoft>
Jun 01 09:04:13 jetson-nano-emmc-cot mender[4877]: time="2021-06-01T09:04:13Z" level=debug msg="Have U-Boot variable: mender_saveenv_canary=1" func=github.com/mendersoftware/mender/in>
Jun 01 09:04:13 jetson-nano-emmc-cot mender[4877]: time="2021-06-01T09:04:13Z" level=debug msg="List of U-Boot variables:map[mender_saveenv_canary:1]" func=github.com/mendersoftware/m>
Jun 01 09:04:13 jetson-nano-emmc-cot mender[4877]: time="2021-06-01T09:04:13Z" level=debug msg="Have U-Boot variable: mender_boot_part=18" func=github.com/mendersoftware/mender/instal>
Jun 01 09:04:13 jetson-nano-emmc-cot mender[4877]: time="2021-06-01T09:04:13Z" level=debug msg="List of U-Boot variables:map[mender_boot_part:18]" func=github.com/mendersoftware/mende>
Jun 01 09:04:13 jetson-nano-emmc-cot mender[4877]: time="2021-06-01T09:04:13Z" level=debug msg="Setting active partition from mount candidate: /dev/mmcblk0p18" func="github.com/mender>
Jun 01 09:04:13 jetson-nano-emmc-cot mender[4877]: time="2021-06-01T09:04:13Z" level=info msg="Mender running on partition: /dev/mmcblk0p18" func="github.com/mendersoftware/mender/cli>
Jun 01 09:04:13 jetson-nano-emmc-cot mender[4877]: time="2021-06-01T09:04:13Z" level=debug msg="Could not open local Mender trust store directory: open /usr/local/share/ca-certificate>
Jun 01 09:04:13 jetson-nano-emmc-cot mender[4877]: time="2021-06-01T09:04:13Z" level=debug msg="ModuleTimeoutSeconds not set. Defaulting to 14400 seconds" func=github.com/mendersoftwa>
Jun 01 09:04:13 jetson-nano-emmc-cot mender[4877]: time="2021-06-01T09:04:13Z" level=info msg="State transition: init [none] -> init [none]" func=github.com/mendersoftware/mender/app.>
Jun 01 09:04:13 jetson-nano-emmc-cot mender[4877]: time="2021-06-01T09:04:13Z" level=debug msg="No state data stored" func="github.com/mendersoftware/mender/app.(*initState).Handle" f>
Jun 01 09:04:13 jetson-nano-emmc-cot mender[4877]: time="2021-06-01T09:04:13Z" level=info msg="State transition: init [none] -> idle [Idle]" func=github.com/mendersoftware/mender/app.>
Jun 01 09:04:13 jetson-nano-emmc-cot mender[4877]: time="2021-06-01T09:04:13Z" level=debug msg="received the GET_AUTH_TOKEN action" func="github.com/mendersoftware/mender/app.(*mender>
Jun 01 09:04:13 jetson-nano-emmc-cot mender[4877]: time="2021-06-01T09:04:13Z" level=info msg="State transition: idle [Idle] -> authorize-wait [Idle]" func=github.com/mendersoftware/m>
Jun 01 09:04:13 jetson-nano-emmc-cot mender[4877]: time="2021-06-01T09:04:13Z" level=debug msg="Handle authorize wait state" func="github.com/mendersoftware/mender/app.(*authorizeWait>
Jun 01 09:04:13 jetson-nano-emmc-cot mender[4877]: time="2021-06-01T09:04:13Z" level=debug msg="Wait 0s before next authorization attempt" func="github.com/mendersoftware/mender/app.(>
Jun 01 09:04:13 jetson-nano-emmc-cot mender[4877]: time="2021-06-01T09:04:13Z" level=info msg="State transition: authorize-wait [Idle] -> authorize [Sync]" func=github.com/mendersoftw>
Jun 01 09:04:13 jetson-nano-emmc-cot mender[4877]: time="2021-06-01T09:04:13Z" level=debug msg="Handle authorize state" func="github.com/mendersoftware/mender/app.(*authorizeState).Ha>
Jun 01 09:04:13 jetson-nano-emmc-cot mender[4877]: time="2021-06-01T09:04:13Z" level=debug msg="received the FETCH_AUTH_TOKEN action" func="github.com/mendersoftware/mender/app.(*mend>
Jun 01 09:04:13 jetson-nano-emmc-cot mender[4877]: time="2021-06-01T09:04:13Z" level=debug msg="Tenant token: dummy" func="github.com/mendersoftware/mender/app.(*menderAuthManagerServ>
Jun 01 09:04:13 jetson-nano-emmc-cot mender[4877]: time="2021-06-01T09:04:13Z" level=debug msg="Authorization data: {{\"mac\":\"00:04:4b:ea:60:c1\"} dummy -----BEGIN PUBLIC KEY-----\n>
Jun 01 09:04:13 jetson-nano-emmc-cot mender[4877]: time="2021-06-01T09:04:13Z" level=debug msg="Making an authorization request () to server https://docker.mender.io" func="github.com>
Jun 01 09:04:13 jetson-nano-emmc-cot mender[4877]: time="2021-06-01T09:04:13Z" level=error msg="Failure occurred while executing authorization request: Method: Post, URL: https://dock>
Jun 01 09:04:13 jetson-nano-emmc-cot mender[4877]: time="2021-06-01T09:04:13Z" level=error msg="Authorize failed: transient error: authorization request failed: transient error: autho>
Jun 01 09:04:13 jetson-nano-emmc-cot mender[4877]: time="2021-06-01T09:04:13Z" level=info msg="State transition: authorize [Sync] -> authorize-wait [Idle]" func=github.com/mendersoftw>
Jun 01 09:04:13 jetson-nano-emmc-cot mender[4877]: time="2021-06-01T09:04:13Z" level=debug msg="Handle authorize wait state" func="github.com/mendersoftware/mender/app.(*authorizeWait>
Jun 01 09:04:13 jetson-nano-emmc-cot mender[4877]: time="2021-06-01T09:04:13Z" level=debug msg="Wait 4m59.864064099s before next authorization attempt" func="github.com/mendersoftware>

#ls -l /var/lib/mender/
total 52
-r--r--r--    1 root     root            40 Mar  9  2018 device_type
-rw-------    1 root     root          2459 Sep 20  2020 mender-agent.pem
-rw-------    1 root     root         32768 Jun  1 09:07 mender-store
-rw-------    1 root     root          8256 Jun  1 09:07 mender-store-lock
-rw-------    1 root     root           199 Jun  1 07:16 mender.conf
drwxr-xr-x    2 root     root          4096 Jun  1 09:04 scripts


Thanks

**Reply by ichergui** [reply]
@tirankas
 
I'm happy with that, it is working fine for you 

*  `$ mender -install` as you can see, your inactive partition was `/dev/mmcblk0p1` and the candidate partition is `/dev/mmcblk0p18` 
* I'm seeing this message `Error: No such container: test_container_del`, it is not related to the A/B system update thoughts. Could you please check that ?

* After rebooting the device, you can apply the commit `$ mender -commit`, this means that there is an update available and commit works fine.

* You cannot see anything happen via `# journalctl -u mender-client` because you are trying a manual A/B system update via CLI and not via a server.

**Note:** you can reboot the device and check the root file system partition (`this is shouldn't changed`).

**Reply by tirankas** [reply]
Hi @ichergui,
No such container is an error I deliberately made as part of the artifactInstall_leave. It means this artifact is executed after install.

My main issue is that I have another artifact, artifactReboot_leavr, which I expected to be executed after the reboot and I don't see it is executed.

Thanks

**Reply by tirankas** [reply]
And forgot to mention that both artifact scripts exist after the update in /var/lib/mender/scripts

**Comment by madisongh** [comment]
Per the [Mender documentation](https://docs.mender.io/artifact-creation/state-scripts#standalone-mode), when you do a "standalone" update (i.e., manually via the `mender install` command), only a small subset of the state scripts get executed.

**Reply by tirankas** [reply]
Thanks @madisongh , indeed...
Working around this by having my systemd service execute the artifact script after reboot and delete it...
Looks good
